In [1]:
# Importing libraries

import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\Ruth Kassaye\Documents\CareerFoundery\DataImmersion\Achievement_4'

In [3]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [4]:
ords_prods_merge.shape

(32404859, 14)

## 1. If-Statements with User-Defined Functions

In [5]:
# Defining a function 
def add_num(x,y):
    sum = x+y
    return sum 

In [6]:
# Run function
add_num(9, 11)

20

###  A user-defined function: sortting products into price ranges
1. If the item’s price is lower than or equal to $5, it will be labeled a “low-range product.”

2. If the item’s price is above USD 5 but lower than or equal to $15, it will be labeled a a “mid-range product.”
3. If the item’s price is above USD 15, it will be labeled a “high-range product.”

In [7]:
# Creating a subset of the ords_prods_merge dataframe as the dataframe is big. 
# Creating subset including everything starting from the beginning of the datafram up to 1 million

df= ords_prods_merge[:1000000]

In [ ]:
df.head()

In [9]:
def price_label(row):
    if row['prices'] <=5:
        return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High range'
    else: return 'Not enough data'

### Designating a location for the 'price_lable' column

In [10]:
# Creating a new column within the df dataframe called “price_range” . axis = 1 tells panda to apply the function to all rows within the dataframe
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\Ruth Kassaye\AppData\Local\Temp\ipykernel_7152\3015330116.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [11]:
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     7.0         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices        price_range  
0         Soda        77              7     9.0  Mid-range product  
1         Soda        77              7     9.0  Mid-range product  
2         Soda        77              7     9.0  Mid-range product  
3         Soda        77              7     9.0  Mid-range product  
4         Soda        77              7     9.0  Mid-range product

In [12]:
# Checking the number of rows and columns
df.shape

(1000000, 15)

In [13]:
# Counting the records in the  price_range column
df['price_range'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

In [14]:
# Filtering out the most expensive product within the df
df['prices'].max()

14.800000190734863

## 2. If-Statements with the loc() Function

- loc() function locates a particular column in the dataframe it’s been assigned to.

In [15]:
# Breaking down the code below called 
# if = df.loc[df['prices'] > 15, then = 'price_range_loc'] = 'High-range product'

df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\Ruth Kassaye\AppData\Local\Temp\ipykernel_7152\4198558753.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [16]:
# Breaking down the code below called 
#if = df.loc[(df['prices'] <= 15) & (df['prices'] > 5), then = 'price_range_loc'] = 'Mid-range product'

df.loc[(df['prices'] <=15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [17]:
# df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'
#df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

df.loc[df['prices'] <=5, 'price_range_loc'] = 'Low-range product'

In [18]:
# Counting the records in the  price_range_loc column
df['price_range_loc'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

## Why loc () better than if statment (explained in section 1)

 - First, using loc() won’t result in a warning message. 
 - Second, the loc() method runs much faster 

### Using loc() function on the ords_prods_merge dataframe

1. If the item’s price is lower than or equal to $5, it will be labeled a “low-range product.”

2. If the item’s price is above USD 5 but lower than or equal to $15, it will be labeled a a “mid-range product.”
3. If the item’s price is above USD 15, it will be labeled a “high-range product.”

In [19]:
ords_prods_merge.loc[ords_prods_merge['prices']>15, 'price_range_loc'] = 'High-range product'

In [20]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <=15) & (ords_prods_merge['prices'] >5), 'price_range_loc'] = 'Mid-range product'

In [21]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [22]:
ords_prods_merge['price_range_loc'].value_counts(dropna= False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

## 3. If-Statements with For-Loops

  - They’re used to perform the same function on multiple elements, for instance, by running through an entire dataframe and performing a function on each row within that dataframe.

In [23]:
for x in range(30,45):
    print('My age is %d' % (x))

My age is 30
My age is 31
My age is 32
My age is 33
My age is 34
My age is 35
My age is 36
My age is 37
My age is 38
My age is 39
My age is 40
My age is 41
My age is 42
My age is 43
My age is 44


In [24]:
# Printing the frequency of the orders_day_of_week to know which day most orders take place
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

#### Analysis of the frequency output above: 
 - 0 (meaning Saturday as per the project brief) has the highest order amount meaning it is the besiest day followed by 1(Friday) and 6(Friday). 4 (Wednesday) is the least busy day.

### Creating columns for 'Busiest day', 'Least busy', 'Regulary busy' using for-loop

In [25]:
# Creating a dataframe using for-loop based in business categories and saving the categories under result
result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [26]:
# Calling result
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [27]:
# Combining the result dataframe with ords_prods_merge dataframe to make sense out of the result dataframe
# This command will combine the result dataframe records in the ords_prods_merge dataframe under a 'busiest day' column

ords_prods_merge['busiest_day'] = result

In [28]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     7.0         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day  
0  Regularly busy  
1  Regularly busy  
2  Regularly busy  
3      Least busy  
4      Least busy

In [29]:
# Printing frequency of the busiest_day column from the ords_prods_merge dataframe
ords_prods_merge['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

## Directions

### Q1. If you haven’t done so already, complete the instructions in the Exercise for creating the “price_label” and “busiest_day” columns.


 - **Answer**: Done. 
      - for the  **'price_lable'** column: please see 'A user-defined function: sortting products into price ranges' under section 1
      - for the **'busiest_day'** column: please see 'Creating columns for 'Busiest day', 'Least busy', 'Regulary busy' using for-loop' under section 3



### Q2. Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

#### First step: Creating 'Busiest days', 'Slowest days' and 'Regularly busy' days label list, named result_2, from the orders_day_of_week column

In [30]:
result_2 = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_2.append("Busiest days")
  elif value == 4 or value == 3:
    result_2.append("Slowest days")
  else:
    result_2.append("Regularly busy")

In [31]:
# Printing result_2 labels
result_2

['Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest 

#### Second step: adding the label list, result_2, in the ords_prods_merge dataframe undre 'busies_day_2' column

In [32]:
# Creating busiest_days column in the ords_prods_merge dataframe

ords_prods_merge['busiest_days'] = result_2

In [33]:
# Calling ords_prods_merge dataframe if the new busiest_day_2 column is added
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     7.0         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day    busiest_days  
0  Regularly busy  Regularly busy  
1  Regularly busy    Slowest days  
2  Regularly busy    Slowest days  
3      Least busy    Slowest days  
4      Least busy    Slowest days

In [35]:
# Printing the frequency of busiest days_2 column of the ords_prods_merge dataframe
ords_prods_merge['busiest_days'].value_counts(dropna = "False")

Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: busiest_days, dtype: int64

In [36]:
# Printing the frequency of order_day_of_week records to check if the total numbers match with my lables I just created above
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

### Q3. Check the values of this new column for accuracy. Note any observations in markdown format.

####  Analysis:
   - As we can see from the frequency result above, overall there are 11,864,412 orders during the two busiest days (sum of 0 and 1 orders). This is also the exact match of Busiest Days label I created above.
   - Also there were 7,624,336 (sum of 3 and 4 orders)orders during the slowest days. This is also the exact match of the Slowest days labels I created above.

### Q4. When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

#### First step: know which hours have the most orders, fewest orders and average ordres using frequncy function

In [37]:
# Printing frequency of busiest order_hours_of_day column to see what are the most orders, fewest orders and average orders
ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

#### Second step: Creating label list - most orders, fewest orders and average ordres from the order_hour_of_day column
 - 10 o'clock would be the hour with the highest orders and 3 o'clock is the hour with the lowest order. Else, would be labeled under average orders

In [38]:
# Creating labels “Most orders,” “Average orders,” and “Fewest orders.”

Busiest_period_of_day = []

for value in ords_prods_merge ['order_hour_of_day']: 
  if value == 10 or value == 11 or value == 14 or value ==15 or value ==13 or value ==12 or value ==16 or value == 9:
    Busiest_period_of_day.append('Most orders')
  elif value == 3 or value ==4 or value ==2 or value ==5 or value ==1 or value ==0 or value ==6 or value ==23:
    Busiest_period_of_day.append('Fewest orders')
  else: 
    Busiest_period_of_day.append('Average orders')

In [39]:
# Calling the Busiest_period_of_day label list created 
Busiest_period_of_day

['Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most ord

#### Third Step: Adding the Busiest_period_of_day label list I created above in the ords_prods_merge dataframe

In [40]:
# Creating a column busiest_period_of_day and assign Busiest_period_of_day label list value to it, see above
ords_prods_merge['busiest_period_of_day'] = Busiest_period_of_day 

In [41]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     7.0         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy        Average orders  
1  Regularly busy    Slowest days        Average orders  
2  Regularly busy    Slowest days           Most orders  
3      Least busy    Slowest days        Average orders  
4      Least busy    Slowest days           Most orders

### Q5. Print the frequency for this new column.

In [42]:
# Frequency of busiest_period_of_day column
ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: busiest_period_of_day, dtype: int64

 - **Analysis**: As we can see above,  the most orders value reflect the orders happened at 10 o'clock and the Fewest orders reflect the ordre happened at 3 o'clock. The rest ordres are labeled as Average ordres 

### Q7. Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.

In [43]:
# Exporting the df_ords_prods_mereg dataframe as pickle format
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_columns_added.pkl'))